subprojects/core/src/main/java/org/gradle/internal/service/scopes/SettingsScopeServices.java

```diff
@@ -54,6 +58,10 @@ public SettingsScopeServices(final ServiceRegistry parent, final SettingsInterna
         });
     }
 
+    protected SettingsLayout createSettingsLayout(FileResolver fileResolver, FileCollectionFactory fileCollectionFactory, FileFactory fileFactory) {
+        return new DefaultSettingsLayout(settings.getSettingsDir(), settings.getRootDir(), fileResolver, fileCollectionFactory, fileFactory);
```

Comment:
🚀 I'd pass the `settings` object instead and call these methods as needed since they might be expensive (`getRootDir` in particular).

Reply:
Ah, because it may lead to walking up the tree to find the nearest settings file?

Reply:
I will look into it, but wonder if keeping a reference to a Settings object may be undesirable. In ProjectLayout, we only keep directory locations.

Reply:
Also, note DefaultSettings.getRootDir() is cached once computed - can't we assume any build will always cause it to be queried (and thus calculated) anyways?